### PR TITLE
custom classification mode/bear detection mode

### PR DESCRIPTION
**Pull request by hexfati**
First implementation of custom classification mode (in the specific case applied to "bear" animal, so "bear detection mode").
The class `CustomClassificationMode` inherits from `AnimalDetectionAndClassificationMode` because shares most of the code with it. The notification is sent only if a "bear" is identified.

### CODE REVIEW

File: wadas/domain/animal_detection_mode.py [hunk 1]
```diff
@@ -46,11 +47,44 @@ def run(self):
...
 
-        self.start_actuator_server()
+    def _detect(self, cur_image_path):
+        results, detected_img_path = self.ai_model.process_image(cur_image_path, True)
+        self.last_detection = detected_img_path
+        return results, detected_img_path
+
+    def _format_classified_animals_string(self, classified_animals):
+        # Prepare a list of classified animals to print in UI
+        if self.last_classified_animals_str:
```
Comment (cdimauro) [inline wadas/domain/animal_detection_mode.py:59]: Isn't this control redundant? I assume that the goal is to initialize / reset the last classification. So, the following line should be enough.

File: wadas/domain/animal_detection_mode.py [hunk 2]
```diff
@@ -46,11 +47,44 @@ def run(self):
...
+            self.last_classified_animals_str = ""
+        for animal in classified_animals:
+            last = animal["classification"][0]
+            if not self.last_classified_animals_str:
+                self.last_classified_animals_str = self.last_classified_animals_str + last
+            else:
+                self.last_classified_animals_str = self.last_classified_animals_str + ", " + last
```
Comment (cdimauro) [inline wadas/domain/animal_detection_mode.py:66]: Maybe you can replace all this code with this (to be checked!):

self.last_classified_animals_str = ", ".join(animal["classification"][0] for animal in classified_animals)

File: wadas/domain/animal_detection_mode.py [hunk 3]
```diff
@@ -46,11 +47,44 @@ def run(self):
...
+
+    def _classify(self, cur_image_path, detection_results):
+        # Classify if detection has identified animals
+        if len(detection_results["detections"].xyxy) > 0:
```
Comment (cdimauro) [inline wadas/domain/animal_detection_mode.py:70]: if len(detection_results["detections"].xyxy):

or (but maybe with pandas or something like that doesn't work. Again, to be checked):

if detection_results["detections"].xyxy:

File: wadas/domain/animal_detection_mode.py [hunk 4]
```diff
@@ -63,44 +97,24 @@ def run(self):
...
+                            cur_img["img"], detected_results
+                        )
+                        if classified_img_path:
                             # Trigger image update in WADAS mainwindow
                             self.update_image.emit(classified_img_path)
                             self.update_info.emit()
-                            message = f"WADAS has classified {self.last_classified_animals}"
-                            message += f"animal from camera {id}!"
+                            message = f"WADAS has classified '{self.last_classified_animals_str}' "
+                            message += f"animal from camera {cur_img['img_id']}!"
```
Comment (cdimauro) [inline wadas/domain/animal_detection_mode.py:117]: Can be combined in one line.

File: wadas/domain/custom_classification_mode.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+                    classified_img_path, classified_animals = self._classify(
+                        cur_img["img"], detected_results
+                    )
+                    if classified_img_path:
+                        # Trigger image update in WADAS mainwindow
+                        self.update_image.emit(classified_img_path)
+                        self.update_info.emit()
+                        message = f"WADAS has classified '{self.last_classified_animals_str}' "
+                        message += f"animal from camera {cur_img['img_id']}!"
```
Comment (cdimauro) [inline wadas/domain/custom_classification_mode.py:54]: Same here.

File: wadas/domain/animal_detection_mode.py [hunk 6]
```diff
@@ -46,11 +47,44 @@ def run(self):
                 ):
                     logger.info("Instantiating FTPS server...")
                     self.ftp_thread = FTPsServer.ftps_server.run()
+        logger.info("Ready for video stream from Camera(s)...")
 
-        self.start_actuator_server()
+    def _detect(self, cur_image_path):
```
Comment (stefanodellosa-personal) [inline wadas/domain/animal_detection_mode.py:52]: I would say that _detect() method could be moved to generic class operation mode, as it is intent of WADAS to perform at least the animal detection.

File: wadas/domain/custom_classification_mode.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                cur_img = None
+
+            if cur_img:
+                logger.debug("Processing image from motion detection notification...")
+                detected_results, detected_img_path = self._detect(cur_img["img"])
+                self.check_for_termination_requests()
+
+                if detected_results and detected_img_path:
+                    # Trigger image update in WADAS mainwindow
+                    self.update_image.emit(detected_img_path)
```
Comment (stefanodellosa-personal) [inline wadas/domain/custom_classification_mode.py:43]: No, we don't update mainwindow with classification detail if classified animal differs from the target. We can only log that the classified animal is different from the target. We can report in log the classified animal but no action is taken in mainwindow.

File: wadas/domain/custom_classification_mode.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                    self.update_info.emit()
+
+                    classified_img_path, classified_animals = self._classify(
+                        cur_img["img"], detected_results
+                    )
+                    if classified_img_path:
```
Comment (stefanodellosa-personal) [inline wadas/domain/custom_classification_mode.py:49]: only if classified is target animal

File: wadas/domain/custom_classification_mode.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+                        if any(
+                            x["classification"][0] == self.target_animal_label
+                            for x in classified_animals
+                        ):
+                            logger.info(message)
+                            self.actuate(cur_img["img_id"])
+                            self.send_notification(classified_img_path, message)
+
+                        else:
+                            logger.debug("Target animal '%s' not found", self.target_animal_label)
```
Comment (stefanodellosa-personal) [inline wadas/domain/custom_classification_mode.py:66]: "Target animal %s not found, skipping notification."

File: wadas/domain/custom_classification_mode.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+                    else:
+                        logger.debug("No results to classify.")
```
Comment (stefanodellosa-personal) [inline wadas/domain/custom_classification_mode.py:69]: "No animal to classify"

File: wadas/domain/custom_classification_mode.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+from wadas.domain.animal_detection_mode import AnimalDetectionAndClassificationMode
+from wadas.domain.camera import img_queue
+
+logger = logging.getLogger(__name__)
+
+
+class CustomClassificationMode(AnimalDetectionAndClassificationMode):
+    """Custom Classification Mode class."""
+
+    def __init__(self, target_animal_label="bear"):
```
Comment (stefanodellosa-personal) [inline wadas/domain/custom_classification_mode.py:15]: We have to check for target animal passed as parameter. it shall match with the list available in the model. (See wadas/ai/models.py list)

**Review by stefanodellosa-personal — CHANGES_REQUESTED** [review]
Please resolve comments before merge

File: wadas/domain/operation_mode.py [hunk 12]
```diff
@@ -66,6 +66,64 @@ def init_model(self):
...
+            for camera in cameras:
+                if camera.type == Camera.CameraTypes.USB_CAMERA:
+                    camera.stop_thread = True
+
+            self.stop_actuator_server()
+
+            self.run_finished.emit()
+            return
+
+    def _initialize_process(self):
```
Comment (stefanodellosa-personal) [inline wadas/domain/operation_mode.py:120]: as we initialize multiple processes, perhaps name should be _initialize_processes()

File: wadas/domain/operation_mode.py [hunk 13]
```diff
@@ -66,6 +66,64 @@ def init_model(self):
...
+        # Initialize ai model
```
Comment (stefanodellosa-personal) [inline wadas/domain/operation_mode.py:121]: we actually do much more than simple Ai model init. please remove comment ad add proper docustring

File: wadas/domain/operation_mode.py [hunk 14]
```diff
@@ -66,6 +66,64 @@ def init_model(self):
         else:
             logger.debug("Model already initialized, skipping initialization.")
 
+    def _initialize_cameras(self):
```
Comment (stefanodellosa-personal) [inline wadas/domain/operation_mode.py:69]: missing docustring

File: wadas/domain/operation_mode.py [hunk 15]
```diff
@@ -66,6 +66,64 @@ def init_model(self):
...
+                elif (
+                    camera.type == Camera.CameraTypes.FTP_CAMERA
+                    and FTPsServer.ftps_server
+                    and not self.ftp_thread
+                ):
+                    logger.info("Instantiating FTPS server...")
+                    self.ftp_thread = FTPsServer.ftps_server.run()
+        logger.info("Ready for video stream from Camera(s)...")
+
+    def _detect(self, cur_image_path):
```
Comment (stefanodellosa-personal) [inline wadas/domain/operation_mode.py:88]: missing docustring

**Review by stefanodellosa-personal — APPROVED** [review]

**Pull request merged**